src/limitations/optional_deps.md

```diff
@@ -0,0 +1,220 @@
+    Feature { base: String, feature: String },
+}
+```
+
+Let's implement the first function required by a dependency provider, `choose_package_version`.
+For that we defined the `base_pkg()` method on a `Package` that returns the string of the base package.
+And we defined the `available_versions()` method on an `Index` to list existing versions of a given package.
+Then we simply called the `choose_package_with_fewest_versions` helper function provided by pubgrub.
+
+```rust
+fn choose_package_version<T: Borrow<Package>, U: Borrow<Range<Version>>>(
+    &self,
+    potential_packages: impl Iterator<Item = (T, U)>,
+) -> Result<(T, Option<Version>), Box<dyn std::error::Error>> {
+    Ok(pubgrub::solver::choose_package_with_fewest_versions(
+        |p| self.available_versions(p.base_pkg()).cloned(),
+        potential_packages,
+    ))
+}
+```
+
+This was very straightforward.
+Implementing the second function, `get_dependencies`, requires just a bit more work but is also quite easy.
+The exact complete version is available in the code, but again, for the sake of simplicity and readability, let's just deal with the happy path.
+
+```rust
+fn get_dependencies(
+    &self,
+    package: &Package,
+    version: &Version,
+) -> Result<Dependencies<Package, Version>, ...> {
+    // Retrieve the dependencies in the index.
+    let deps =
+        self.packages
+            .get(package.base_pkg()).unwrap()
+            .get(version).unwrap();
+
+    match package {
+        // If we asked for a base package, we simply return the mandatory dependencies.
+        Package::Base(_) => Ok(Dependencies::Known(from_deps(&deps.mandatory))),
+        // Otherwise, we concatenate the feature deps with a dependency to the base package.
+        Package::Feature { base, feature } => {
+            let feature_deps = deps.optional.get(feature).unwrap();
+            let mut all_deps = from_deps(feature_deps);
+            all_deps.insert(
+                Package::Base(base.to_string()),
+                Range::exact(version.clone()),
+            );
+            Ok(Dependencies::Known(all_deps))
+        },
+    }
+}
+```
+
+Quite easy right?
+The helper function `from_deps` is where each dependency is transformed from its `String` form into its `Package` form.
+In pseudo-code (not compiling, but you get how it works) it looks like follows.
+
+```rust
+/// Helper function to convert Index deps into what is expected by the dependency provider.
+fn from_deps(deps: &Map<String, Dep>) -> DependencyConstraints<Package, Version> {
+    deps.iter().map(create_feature_deps)
+        .flatten()
+        .collect()
+}
+
+fn create_feature_deps(
+    (base_pkg, dep): (&String, &Dep)
+) -> impl Iterator<Item = (Package, Range<Version>)> {
+    if dep.features.is_empty() {
+        // If the dependency has no optional feature activated,
+        // we simply add a dependency to the base package.
+        std::iter::once((Package::Base(base_pkg), dep.range))
```

Comment:
Cant this just return and empty Iterator, like `None.into_iter()`?

Reply:
I don't think so. I'll double check.

Reply:
The dependency to the base package is indeed required with how I setup things.

Reply:
Then why is it not needed in the `else` case?

Reply:
Because each feature package has itself a dependency to the base package.

Reply:
We could also add the dependency in both branches of this but it wasteful since each feature package has a dependency to the base package too.

Reply:
By the way, the exact code is slightly different from what I put here in the doc. The actual code is this:
```rust
/// Helper function to convert Index deps into what is expected by the dependency provider.
fn from_deps(deps: &Map<String, Dep>) -> DependencyConstraints<Package, Version> {
    deps.iter()
        .flat_map(|(base_pkg, dep)| {
            let feature_count = dep.features.len();
            dep.features
                .iter()
                .map(move |feat| {
                    (
                        Package::Feature {
                            base: base_pkg.clone(),
                            feature: feat.clone(),
                        },
                        dep.range.clone(),
                    )
                })
                .chain(std::iter::once((
                    Package::Base(base_pkg.clone()),
                    dep.range.clone(),
                )))
                // If there was no feature, we take the base package, otherwise, we don't.
                .take(feature_count.max(1))
        })
        .collect()
```

Reply:
Ah yes, I am miss reading sorry.

Reply:
No problem at all!